src/Extensions/PdoExecuteTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,208 @@
+        if (null === $queryExpr) {
+            return null;
+        }
+
+        // resolve query parameter from "prepare"
+        if ($queryExpr instanceof MethodCall) {
+            $args = $queryExpr->getArgs();
+            $queryExpr = $args[0]->value;
+        }
+
+        $queryReflection = new QueryReflection();
+        $queryString = $queryReflection->resolveQueryString($queryExpr, $scope);
+        if (null === $queryString) {
+            return null;
+        }
+
+        foreach ($placeholders as $placeholderName => $value) {
+            $queryString = str_replace($placeholderName, $value, $queryString);
+        }
+        $reflectionFetchType = QueryReflector::FETCH_TYPE_BOTH;
+        $resultType = $queryReflection->getResultType($queryString, $reflectionFetchType);
+
+        if ($resultType) {
+            return new GenericObjectType(PDOStatement::class, [$resultType]);
+        }
+
+        return null;
+    }
+
+    /**
+     * @return array<string, string>
+     */
+    public function resolveParameters(Type $parameterTypes): array
+    {
+        $placeholders = [];
+
+        if ($parameterTypes instanceof ConstantArrayType) {
+            $keyTypes = $parameterTypes->getKeyTypes();
+            $valueTypes = $parameterTypes->getValueTypes();
+
+            foreach ($keyTypes as $i => $keyType) {
+                if ($keyType instanceof ConstantStringType) {
+                    $placeholderName = $keyType->getValue();
+
+                    if (!str_starts_with($placeholderName, ':')) {
+                        $placeholderName = ':'.$placeholderName;
+                    }
+
+                    if ($valueTypes[$i] instanceof ConstantScalarType) {
+                        $placeholders[$placeholderName] = (string) ($valueTypes[$i]->getValue());
+                    }
+                }
+            }
+        }
+
+        return $placeholders;
+    }
+
+    private function findQueryStringExpression(MethodCall $methodCall): ?Expr
+    {
+        // todo: use astral simpleNameResolver
+        $nameResolver = function ($node) {
+            if (\is_string($node->name)) {
+                return $node->name;
+            }
+            if ($node->name instanceof Node\Identifier) {
+                return $node->name->toString();
+            }
+        };
+
+        $current = $methodCall;
+        while (null !== $current) {
+            /** @var Assign|null $assign */
+            $assign = $this->findFirstPreviousOfNode($current, function ($node) {
+                return $node instanceof Assign;
+            });
+
+            if (null !== $assign && $nameResolver($assign->var) === $nameResolver($methodCall->var)) {
+                return $assign->expr;
+            }
+
+            $current = $assign;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param callable(Node $node):bool $filter
+     */
+    public function findFirstPreviousOfNode(Node $node, callable $filter): ?Node
```

Comment:
will hopefully be used from astral in the near fuure

https://github.com/symplify/symplify/pull/3876

Reply:
https://github.com/staabm/phpstan-dba/issues/51